### ISSUE TITLE
502 error if a footnote call points to a missing footnote

### ISSUE THREAD

**Issue by aleray** [torchbox/wagtail-footnotes#11]
Hello,

I get 502 error if a footnote call points to a missing footnote. It's was kind of hard to debug. Would there be a way to raise a validation issue in the admin in case a footnote call refers to a missing footnote?

Thanks

**Comment by zerolab** [comment]
@aleray, would you be able to submit a full Django error stack trace?

**Comment by aleray** [comment]
Sure,

here it is (not so hard to debug DEBUG mode, but it was in prod under Docker):

```
Request Method: GET
Request URL: http://medor.127.0.0.1.nip.io:8000/nos-questions/la-belgique-va-t-elle-bousiller-son-patrimoine-cine/episodes/cinematek-piratage-et-retour-a-la-clandestinite/

Django Version: 2.2.17
Python Version: 3.7.9
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.sitemaps',
 'wagtail.api.v2',
 'wagtail.contrib.forms',
 'wagtail.contrib.modeladmin',
 'wagtail.contrib.postgres_search',
 'wagtail.contrib.redirects',
 'wagtail.contrib.settings',
 'wagtail.embeds',
 'wagtail.sites',
 'wagtail.users',
 'wagtail.snippets',
 'wagtail.documents',
 'wagtail.images',
 'wagtail.search',
 'wagtail.admin',
 'wagtail.core',
 'wagtail.contrib.routable_page',
 'modelcluster',
 'taggit',
 'rest_framework',
 'rest_framework.authtoken',
 'corsheaders',
 'django_extensions',
 'wagtailfontawesome',
 'wagtailmenus',
 'social_django',
 'wagtailstreamforms',
 'django_rq',
 'wagtailautocomplete',
 'wagtail_footnotes',
 'anymail',
 'honeypot',
 'medor_wagtail_images.apps.MedorWagtailImagesConfig',
 'medor_article.apps.MedorArticleConfig',
 'medor_auth.apps.MedorAuthConfig',
 'medor_core.apps.MedorCoreConfig',
 'medor_menus.apps.MedorMenusConfig',
 'medor_home.apps.MedorHomeConfig',
 'medor_people.apps.MedorPeopleConfig',
 'medor_mag.apps.MedorMagConfig',
 'medor_faq.apps.MedorFaqConfig',
 'medor_thematic.apps.MedorThematicConfig',
 'medor_tour.apps.MedorTourConfig',
 'medor_hypersurveillance.apps.MedorHypersurveillanceConfig',
 'medor_search',
 'medor_simple_page.apps.MedorSimplePageConfig',
 'medor_account.apps.MedorAccountConfig',
 'medor_events.apps.MedorEventsConfig',
 'medor_forms.apps.MedorFormsConfig',
 'medor_newsletter.apps.MedorNewsletterConfig',
 'medor_quiz.apps.MedorQuizConfig',
 'medor',
 'debug_toolbar']
Installed Middleware:
['corsheaders.middleware.CorsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'medor_newsletter.middleware.CustomCsrfViewMiddleware',
 'medor.middleware.CustomHoneypotMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'wagtail.contrib.redirects.middleware.RedirectMiddleware',
 'medor_auth.middleware.UpdateOdooDataMiddleware',
 'medor_newsletter.middleware.AddTOCResponseMiddleware',
 'medor_newsletter.middleware.PremailerResponseMiddleware',
 'debug_toolbar.middleware.DebugToolbarMiddleware']


Template error:
In template /app/medor/templates/organisms/section_article.html, error at line 90
   Footnote matching query does not exist.
   80 : 
   81 : 
   82 :       <section class="cb cb-text cb-text--intro">
   83 :         {{ self.header_text|medor_richtext|remove_links }}
   84 :       </section>
   85 : 
   86 : 
   87 :       <span id="continuer-a-lire"></span>
   88 : 
   89 :       {% for block in page.content %}
   90 :          {% include_block block %} 
   91 :       {% endfor %}
   92 : 
   93 :       {% if self.tags.all %}
   94 :       {% include "molecules/cb_tags.html" with article=self %}
   95 :       {% endif %}
   96 : 
   97 :       <footer class="article__footer">
   98 :         {% if self.is_serie_article_item %}
   99 :           {% if self.prev_article or self.next_article %}
   100 :             {% include 'molecules/content_nav.html' with prev_url=self.prev_article.get_url prev_text="Épisode précédent" next_url=self.next_article.get_url next_text="Épisode suivant" %}


Traceback:

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  145.                 response = self.process_exception_by_middleware(e, request)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/core/handlers/base.py" in _get_response
  143.                 response = response.render()

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/response.py" in render
  106.             self.content = self.rendered_content

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/response.py" in rendered_content
  83.         content = template.render(context, self._request)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/backends/django.py" in render
  61.             return self.template.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  171.                     return self._render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/test/utils.py" in instrumented_test_render
  96.     return self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/test/utils.py" in instrumented_test_render
  96.     return self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  150.             return compiled_parent._render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/test/utils.py" in instrumented_test_render
  96.     return self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/templatetags/tz.py" in render
  109.             output = self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/defaulttags.py" in render
  109.         output = self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  62.                 result = block.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/loader_tags.py" in render
  188.             return template.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  173.                 return self._render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/test/utils.py" in instrumented_test_render
  96.     return self.nodelist.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render
  937.                 bit = node.render_annotated(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/defaulttags.py" in render
  209.                     nodelist.append(node.render_annotated(context))

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/template/base.py" in render_annotated
  904.             return self.render(context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/wagtail/core/templatetags/wagtailcore_tags.py" in render
  132.             return value.render_as_block(context=new_context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/wagtail/core/blocks/base.py" in render_as_block
  438.         return self.block.render(self.value, context=context)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/wagtail_footnotes/blocks.py" in render
  34.         return self.replace_footnote_tags(html, context=context)

File "/app/medor_core/blocks.py" in replace_footnote_tags
  32.         return mark_safe(FIND_FOOTNOTE_TAG.sub(replace_tag, html))

File "/app/medor_core/blocks.py" in replace_tag
  25.                 index = self.process_footnote(match.group(1), context["page"])

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/wagtail_footnotes/blocks.py" in process_footnote
  43.         footnote = page.footnotes.get(uuid=footnote_id)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/db/models/manager.py" in manager_method
  82.                 return getattr(self.get_queryset(), name)(*args, **kwargs)

File "/home/medor/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/django/db/models/query.py" in get
  408.                 self.model._meta.object_name

Exception Type: DoesNotExist at /nos-questions/la-belgique-va-t-elle-bousiller-son-patrimoine-cine/episodes/cinematek-piratage-et-retour-a-la-clandestinite/
Exception Value: Footnote matching query does not exist.
```

**Comment by zerolab** [comment]
Fantastic, thank you.

**Comment by jafacakes2011** [comment]
This MR should fix the immediate 502 error #16 

I have a TODO on that Pull Request to add validation to prevent saving a Page with invalid Footnote references, but I am not 100% sure how to approach that at the moment. I'd be open to some comments/suggestions.
